feat: create Code Block Package (#2641)
twilio-labs/paste: packages/paste-website/src/pages/components/code-block/index.mdx

export const pageQuery = graphql`
  {
    allPasteComponent(filter: {name: {eq: "@twilio-paste/code-block"}}) {
      edges {
        node {
          name
          description
          status
          version
        }
      }
    }
    mdx(frontmatter: {slug: {eq: "/components/code-block/"}}) {
      fileAbsolutePath
      frontmatter {
        slug
        title
        description
      }
      headings {
        depth
        value
      }
    }
    allAirtable(filter: {data: {Feature: {eq: "Code Block"}}}) {
      edges {
        node {
          data {
            Documentation
            Figma
            Design_committee_review
            Engineer_committee_review
            Code
            status
          }
        }
      }
    }
  }
`;

<NormalizedComponentHeader
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/code-block"
  storybookUrl="/?path=/story/components-code-block--default"
  data={props.data}
  description={props.data.mdx.frontmatter.description}
/>

---

<contentwrapper>

<PageAside data={props.data.mdx} />

<content>

> 

Author: Note: once the multi line example works, I will add it to the top of the page

## Guidelines

### About Code Block

The Code Block component improves readability of embedded code samples with syntax highlighting and automatic line numbering.

Multiple Code Blocks may be combined with the Code Block Group component. A common use case for Code Block Group is providing a code sample in multiple languages.

### Accessibility

Both Code Block and Code Block Group are fully keyboard accessible. Here are some tips to ensure a positive user experience for all users:

> - Provide a descriptive label such as the current filename or step in a process to clarify the purpose of the Snippet.

Reviewer: nit: a wild Snippet reference appears

- Enable long line wrapping whenever possible for a better reading experience.

## Examples

### Single line

Single line Snippets exclude the label bar and line numbers. It is automatically detected but will default to the multi line layout if line wrapping is enabled.

<LivePreview scope={{CodeBlock}} language="jsx">
  {`<CodeBlock language='shell-session'>
  ${`$ curl -G https://api.twilio.com/2010-04-01/Accounts -u '[YOUR ACCOUNT SID]:[YOUR AUTH TOKEN]'`}
</CodeBlock>`}
</LivePreview>

### Multi line

A multi line snippet displays a label and line numbers by default. If no label is provided the snippet language will be used as a fallback.

<LivePreview scope={{CodeBlock}} language="jsx" noInline>
  {multiLine}
</LivePreview>

> ### Grouped

Author: Note: I'm waiting to add the grouped example until I figure out the multi line example.

### Accessibility

To ensure the Code Block is accessible:

- Provide a descriptive label such as the current filename or step in a process to clarify the purpose of the Code Block.
- Enable long line wrapping whenever possible for a better reading experience.
- Use the correct heading level for the CodeBlockHeader

## Examples

### Single line

Use `variant="single-line"` for one line Code Blocks. Single line Code Blocks should not have `showLineNumbers={true}`.

<LivePreview scope={{CodeBlock}} language="jsx" noInline>
  {singleLine}
</LivePreview>

### Multi line

Code Blocks have `variant="multi-line"` by default. Multi line Code Blocks should have a CodeBlockHeader with a logical label for the Code Block, like the language or file name.

Note for the `maxLines` property: Code Block uses [line-clamp](https://developer.mozilla.org/en-US/docs/Web/CSS/-webkit-line-clamp) to limit the number of lines visible. When using line-clamp, browsers add an ellipses to the contents of the Code Block to indicate that there is more to scroll to.

<LivePreview scope={{CodeBlockWrapper, CodeBlockHeader, CodeBlock}} language="jsx" noInline>
  {multiLine}
</LivePreview>

### Tabs

> Multiple Code Blocks may displayed as a tabset with the CodeBlockTabList, CodeBlockTab, and CodeBlockTabPanel components. Provide a logical name for each CodeBlockTab, like the language or file name.

Reviewer 2: ```suggestion
Multiple Code Blocks may be displayed as a tabset with the CodeBlockTabList, CodeBlockTab, and CodeBlockTabPanel components. Provide a logical name for each CodeBlockTab, like the language or file name.
```

##### CodeBlockTab

| Prop       | Type     | Description                                                                               | Default            |
| ---------- | -------- | ----------------------------------------------------------------------------------------- | ------------------ |
| `children` | `string` | The code snippet to be rendered.                                                          |                    |
| `element?` | `string` | Overrides the default element name to apply unique styles with the Customization Provider | `'CODE_BLOCK_TAB'` |

##### CodeBlockTabPanel

| Prop       | Type     | Description                                                                               | Default                  |
| ---------- | -------- | ----------------------------------------------------------------------------------------- | ------------------------ |
| `children` | `string` | The code snippet to be rendered.                                                          |                          |
| `element?` | `string` | Overrides the default element name to apply unique styles with the Customization Provider | `'CODE_BLOCK_TAB_PANEL'` |

> ##### CodeBlock

Reviewer 3: suggestion: I think this prop table should go first or second, since it's probably the one people will be referring to most.

Author: Good point!

| `variant?`             | `'single-line', 'multi-line'`                                                                                           | Which variant of code block to render.                                                    | `'multi-line'`                  |
| `showLineNumbers?`     | `boolean`                                                                                                               | Include line numbers in code block.                                                       | `false`                         |
| `wrapLines?`           | `boolean`                                                                                                               | Wraps long lines of code.                                                                 | `false`                         |
| `maxLines?`            | `boolean`                                                                                                               | Constrains the height of code block to a set number of lines.                             |                                 |

> | `externalLink?`        | `string`                                                                                                                | Link to the full source code.                                                             |                                 |

Reviewer 3: ```suggestion
| `externalLink?`        | `string`                                                                                                                | Optional link to an external source for the code                                                            |                                 |
```

> | `i18nCopyLabelBefore?` | `string`                                                                                                                | The content of the copy button tooltip.                                                   | `'Copy code block'`             |

Reviewer 3: ```suggestion
| `i18nCopyLabelBefore?` | `string`                                                                                                                | The content of the copy button tooltip before it is clicked.                                                   | `'Copy code block'`             |
```

> | `i18nCopyLabelAfter?`  | `string`                                                                                                                | The content of the copy button tooltip when the user clicks it.                           | `'Copied!'`                     |

Reviewer 3: ```suggestion
| `i18nCopyLabelAfter?`  | `string`                                                                                                                | The content of the copy button tooltip after the user clicks it.                           | `'Copied!'`                     |
```

> | `i18nLinkLabel?`       | `string`                                                                                                                | The content of the external link tooltip.                                                 | `'Open code block in new page'` |

Reviewer 3: ```suggestion
| `i18nLinkLabel?`       | `string`                                                                                                                | The content of the external link button tooltip.                                                 | `'Open code block in new page'` |
```
